File: src/ui/web/chart_bridge.py
# File: src/ui/web/chart_bridge.py
"""The Python half of the ``chart_host.html`` <-> Qt event channel.

Two directions, and only one needs this class:

- **Python -> JS** (pushing a new figure, or a theme's relayout patch) goes
  straight through ``QWebEnginePage.runJavaScript()`` from
  :class:`~src.ui.widgets.chart_view.ChartView` -- a one-way call needs no
  registered object on the other end.
- **JS -> Python** (a Plotly click or box/lasso selection) needs a
  :class:`QObject` :class:`QWebChannel` can expose to the page, which is
  what :class:`ChartBridge` is. ``resources/web/chart_bridge.js`` looks it
  up as ``channel.objects.chartBridge`` and calls its ``@Slot`` methods.

:attr:`ChartBridge.point_clicked` and :attr:`ChartBridge.selection_changed`
have no consumer yet -- milestone 24 uses them to filter a paired
``DataTableView`` from a chart click/selection. Defining the channel
contract now, even though nothing reads it yet, means
``resources/web/chart_bridge.js`` and ``chart_host.html`` do not need to
change again when M24 lands, matching how milestone 15 built
:mod:`src.ui.theme.plotly_theme` unused until this very milestone wired it
into chart rendering.
"""

from __future__ import annotations

import json

from PySide6.QtCore import QObject, Signal, Slot

from src.core.logger import get_logger

_logger = get_logger(__name__)


class ChartBridge(QObject):
    """Registered as ``"chartBridge"`` on a chart view's :class:`QWebChannel`.

    Args:
        parent: Optional owning :class:`QObject` -- typically the
            :class:`~src.ui.widgets.chart_view.ChartView` that constructs
            it, so the bridge's lifetime matches the view's.

    Signals:
        point_clicked: Emitted with a ``dict`` (``curveNumber``,
            ``pointIndex``, ``x``, ``y``) when a single Plotly marker is
            clicked.
        selection_changed: Emitted with a ``list[int]`` of point indices
            when a box/lasso selection completes.
    """

    point_clicked = Signal(dict)
    selection_changed = Signal(list)

    @Slot(str)
    def notify_point_clicked(self, payload_json: str) -> None:
        """Invoked from JS (``forwardClick`` in ``chart_bridge.js``).

        Args:
            payload_json: A JSON object string, e.g.
                ``'{"curveNumber": 0, "pointIndex": 3, "x": 1.5, "y": 22.0}'``.
                Malformed input is logged and dropped rather than raised --
                this method executes inside a Qt slot invoked from the
                WebEngine render process, where an uncaught exception has no
                well-defined place to surface to the user.
        """
        try:
            payload = json.loads(payload_json)
        except (TypeError, ValueError):
            _logger.warning(
                "ChartBridge received a malformed point-click payload: %r",
                payload_json,
            )
            return
        if not isinstance(payload, dict):
            _logger.warning(
                "ChartBridge point-click payload was not a JSON object: %r",
                payload,
            )
            return
        self.point_clicked.emit(payload)

    @Slot(str)
    def notify_selection_changed(self, payload_json: str) -> None:
        """Invoked from JS (``forwardSelection`` in ``chart_bridge.js``).

        Args:
            payload_json: A JSON array-of-integers string, e.g. ``"[3, 4, 5]"``.
        """
        try:
            payload = json.loads(payload_json)
        except (TypeError, ValueError):
            _logger.warning(
                "ChartBridge received a malformed selection payload: %r",
                payload_json,
            )
            return
        if not isinstance(payload, list):
            _logger.warning(
                "ChartBridge selection payload was not a JSON array: %r",
                payload,
            )
            return
        self.selection_changed.emit(payload)

File: resources/web/chart_bridge.js
// chart_bridge.js -- the JS half of the chart_view.py <-> chart_host.html
// channel. Exposes renderFigure()/relayout(), called from Python via
// QWebEngineView.page().runJavaScript(), and forwards Plotly click/selection
// events back to Python through the "chartBridge" QWebChannel object (see
// src/ui/web/chart_bridge.py::ChartBridge).
//
// Click/selection forwarding has no Python-side consumer yet -- that lands
// in milestone 24, which filters a paired DataTableView from a chart click.
// The channel contract is defined now so chart_host.html does not need to
// change again then, matching how src/ui/theme/plotly_theme.py was built in
// milestone 15 and only wired into chart rendering here in milestone 16.
(function () {
  "use strict";

  var chartEl = document.getElementById("chart");
  var bridge = null;
  var rendered = false;

  function connectBridge() {
    // qt.webChannelTransport only exists inside a QWebEngineView that has
    // called page().setWebChannel(...) -- absent when this file is opened
    // as a plain local file (e.g. during manual debugging), in which case
    // rendering still works and forwarding is silently a no-op.
    if (typeof qt === "undefined" || !qt.webChannelTransport) {
      return;
    }
    new QWebChannel(qt.webChannelTransport, function (channel) {
      bridge = channel.objects.chartBridge || null;
    });
  }

  // Called from Python with three JSON-encoded strings (data, layout,
  // config) rather than one combined object -- layout is assembled
  // Python-side by merging the figure's own layout with
  // src.ui.theme.plotly_theme.plotly_layout(tokens), and keeping the pieces
  // separate here mirrors that split instead of re-merging it in JS too.
  window.renderFigure = function (dataJson, layoutJson, configJson) {
    var data = JSON.parse(dataJson);
    var layout = JSON.parse(layoutJson);
    var config = JSON.parse(configJson);
    if (!rendered) {
      Plotly.newPlot(chartEl, data, layout, config);
      rendered = true;
      chartEl.on("plotly_click", forwardClick);
      chartEl.on("plotly_selected", forwardSelection);
    } else {
      // Plotly.react diffs against the existing plot rather than
      // rebuilding the DOM from scratch -- this is what makes a re-render
      // (e.g. after a data update) not flicker the way a fresh newPlot()
      // would.
      Plotly.react(chartEl, data, layout, config);
    }
  };

  // Called on every theme toggle (src/ui/dock_manager.py, subscribed to
  // ThemeManager.theme_changed) with just the new layout -- Plotly.relayout
  // patches colours/fonts in place with no re-plot of the data traces
  // themselves, which is what keeps a theme switch from flickering.
  window.relayout = function (layoutJson) {
    if (!rendered) {
      return;
    }
    Plotly.relayout(chartEl, JSON.parse(layoutJson));
  };

  function forwardClick(evt) {
    if (!bridge || !evt || !evt.points || !evt.points.length) {
      return;
    }
    var point = evt.points[0];
    bridge.notify_point_clicked(
      JSON.stringify({
        curveNumber: point.curveNumber,
        pointIndex: point.pointIndex,
        x: point.x,
        y: point.y,
      })
    );
  }

  function forwardSelection(evt) {
    if (!bridge || !evt || !evt.points) {
      return;
    }
    var indices = evt.points.map(function (point) {
      return point.pointIndex;
    });
    bridge.notify_selection_changed(JSON.stringify(indices));
  }

  connectBridge();
})();
